src/arch/arm/object/vcpu.c

```diff
@@ -67,6 +67,14 @@ static word_t readVCPUReg(vcpu_t *vcpu, word_t field)
             } else {
                 return vcpu_read_reg(vcpu, field);
             }
+#ifdef CONFIG_HAVE_FPU
+        case seL4_VCPUReg_CPACR:
+            if (ARCH_NODE_STATE(armHSVCPUActive)) {
```

Comment:
Does this only check if there is a VCPU active _at all_ rather than if the active VCPU is the invoked VCPU?

Reply:
i.e. if there is another VCPU active, could this invocation end up modifying that VCPU instead of the one for which the capability was invoked?

Reply:
@Xaphiosis do we have an invariant that `vcpu` must be the active VCPU at this stage? Because if not, then `SCTLR` also seems broken by the same argument (which if we don't have that invariant would be possible).

Reply:
Unfortunately, from memory we were halfway through proving that invariant before we had to change focus to a different project.

Reply:
In the Haskell, we do what the C does here and only do anything if the supplied vcpu is the active vcpu. Otherwise, we do nothing, regardless of whose VCPU we were given.
We are still missing an invariant (proofs got halfway, then resources were needed elsewhere) that the active VCPU is the current thread's VCPU (checked with Corey). I'm reasonably sure it doesn't happen, but verification does not guarantee it.
Also, currently ARM_HYP proofs are for FPU disabled, which means we won't see this, but the SCTLR issue remains.

Reply:
> Does this only check if there is a VCPU active _at all_ rather than if the active VCPU is the invoked VCPU?

To be honest, I'm not sure. I made that *assumption* based on the SCTRL implementation (which you can see if you expand the diff out a bit). If that assumption is actually wrong then SCTRL is also broken!

I would like to get this merged irrespective of that assumption at this point. If that assumption is indeed wrong, I think we should get this merged and separately fix both SCTRL and CPACR at the same (if only for the point that I've hit the CPACR problem in real code and it is a blocker for me, but I've not hit the theoretical SCTRL problem yet, and it isn't blocking me).

Reply:
@bennoleslie Could you share what the failed case looks like?

Reply:
The failure is user code crashes when trying to write a value to the CPACR register of a VCPU. (Given that the actual CPACR register just got changed underneath it this is not surprising).

If you never write a CPACR register explicitly you aren't going to see the problem.

Reply:
Ok, I think this is Ok to merge, because it is at least no worse than `SCTRL`, but we should put that VCPU invariant back on our radar.

Reply:
> Does this only check if there is a VCPU active _at all_ rather than if the active VCPU is the invoked VCPU?

There is a check above, armHSCurVCPU == vcpu. This returns last-saved registers for a cross-core read, if the VCPU register read and the VCPU are on different cores. This is another issue.